File: mvvm/src/main/java/com/lee/mvvm/base/view/widge/rv/BaseAdapter.java
package com.lee.mvvm.base.view.widge.rv;

import android.databinding.DataBindingUtil;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

/**
 * this adapter is use for DataBinding&RecycleView
 *
 * @param <T> data type
 */
public class BaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder>
        implements View.OnClickListener {

    private OnItemClickListener onItemClickListener;

    protected int layoutId;
    protected int variableId;
    protected List<T> data;

    /**
     * @param layoutId   item layout id
     * @param variableId variable id: the name of variable
     * @param data       data
     */
    public BaseAdapter(@LayoutRes int layoutId, int variableId, @NonNull List<T> data) {
        this.layoutId = layoutId;
        this.variableId = variableId;
        this.data = data;
    }

    @NonNull
    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = DataBindingUtil
                .inflate(LayoutInflater.from(viewGroup.getContext()), layoutId, null, false)
                .getRoot();
        v.setTag(i);
        v.setOnClickListener(this);
        return new BaseViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder baseViewHolder, int i) {
        convert(baseViewHolder, data.get(i));
    }

    @Override
    public void onClick(View v) {
        if (onItemClickListener != null) {
            onItemClickListener.onItemClick((Integer) v.getTag());
        }
    }

    public void convert(BaseViewHolder holder, T t) {
        holder.setBinding(variableId, t);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public interface OnItemClickListener {
        void onItemClick(int position);
    }
}

File: mvvm/src/main/java/com/lee/mvvm/base/view/widge/rv/BaseAdapter.kt
package com.lee.mvvm.base.view.widge.rv

import androidx.databinding.DataBindingUtil
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.ViewDataBinding

/**
 * this adapter is use for DataBinding&RecycleView
 *
 * @param <T> data type
</T> */
open class BaseAdapter<T : Any>
(@param:LayoutRes var layoutId: Int, var variableId: Int, var data: List<T>) :
        RecyclerView.Adapter<BaseViewHolder<*>>(), View.OnClickListener {

    private var onItemClickListener: OnItemClickListener? = null

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): BaseViewHolder<*> {
        val v = DataBindingUtil
                .inflate<ViewDataBinding>(LayoutInflater.from(viewGroup.context), layoutId, null, false)
                .root
        v.tag = i
        v.setOnClickListener(this)
        return BaseViewHolder<ViewDataBinding>(v)
    }

    override fun onBindViewHolder(baseViewHolder: BaseViewHolder<*>, i: Int) {
        convert(baseViewHolder, data[i])
    }

    override fun onClick(v: View) {
        if (onItemClickListener != null) {
            onItemClickListener!!.onItemClick(v.tag as Int)
        }
    }

    private fun convert(holder: BaseViewHolder<*>, t: T) {
        holder.setBinding(variableId, t)
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener) {
        this.onItemClickListener = onItemClickListener
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int)
    }
}
